conda/models/match_spec.py

```diff
@@ -842,6 +842,44 @@ def exact_value(self):
     def matches_all(self):
         return self._raw_value == '*'
 
+    def merge(self, other):
+        # exact match
+        if self.raw_value == other.raw_value:
+            return self.raw_value
+
+        if not self._re_match and isinstance(other, GlobStrMatch) and other._re_match:
+            # swap order, so `self` always has an actual pattern if there is only one
+            other, self = self, other
+
+        # the other component might not have str 'raw_value' (e.g. Channel or MultiChannel)
+        other_as_str = str(other)
+
+        if "*" not in other_as_str:
+            # other is an exact literal,
+            # check our pattern against it
+            # if we match, other is more strict
+            if self._re_match and self._re_match(other_as_str):
+                return other.raw_value
+            else:
+                # Raise on incompatible pattern
+                return super().merge(other)
+
+        # both are patterns, compute regular expression intersection
+        # compatibility to be determined by the index filters
+        patterns = []
+        for m in (self, other):
+            value = str(m)
+            if value.startswith("^") and value.endswith("$"):
+                patterns.append(value[1:-1])
+            elif "*" in value:
+                value = re.escape(value).replace(r"\*", r".*")
+                patterns.append(value)
+            else:
+                patterns.append(value)
+
+        # lookahead assertion followed by non-capture group
+        return rf"^(?={patterns[0]})(?:{patterns[1]})$"
```

Comment:
Note that the resulting intersection regex cannot be parsed correctly by the `libmamba` solver right now. I submitted an issue: https://github.com/mamba-org/mamba/issues/1853

I tried coming up with a glob-only solution, but there are too many edge cases the more you think about it.

Reply:
I _also_ tried coming up with a glob solution! I think it would be better if it could be done, but I wasn't confident I could accomplish it in general. I'm also not sure there _is_ always a single glob to represent all mutual matches of two globs, even though the reduced case of only considering wildcards at the start and/or end simplifies things greatly, is solvable, and probably covers ~all cases in reality.

Reply:
Even if we add a condition set for "no wildcards in the middle of the string", there are too many combinations, I think.

Definitions:
- `needle*` as "prefix glob"
- `*needle*` as "infix glob"
- `*needle` as "suffix glob"
- `*ne*dle*` as "regex-required glob", which has four variants (with and/or without leading/trailing globs)

1. `prefix ^ prefix`: only compatible if one of the prefixes is a substring of the other, we keep the longest prefix. `nee*` ^ `needle*` = `needle*`
2. `prefix ^ infix`: inner-left merge. `nee*` ^ `*dle*` = `nee*dle*`
3. `prefix ^ suffix`: inner-inner merge. `nee*` ^ `*dle` = `nee*dle`
4. `prefix ^ regex`: regex intersection needed
5. `infix ^ prefix`: another case of (2)
6. `infix ^ infix`: regex intersection needed, since solution is order dependent
7. `infix ^ suffix`: another case of (2)
8. `infix ^ regex`: regex intersection needed
9. `suffix ^ prefix`: same as (3)
10. `suffix ^ infix`: same as (7)
11. `suffix ^ suffix`: same as (1)
12. `suffix ^ regex`: regex intersection needed
13. `regex ^ prefix`: regex intersection needed
14. `regex ^ infix`: regex intersection needed
15. `regex ^ suffix`: regex intersection needed
16. `regex ^ regex`: regex intersection needed

So simplifying, we have three strategies:

1. Accept the longest if the literals are compatible, reject otherwise (valid for prefix^prefix or suffix^suffix)
2. Merge by inner stars (valid for prefix^infix, infix^suffix or prefix^suffix)
3. Use regex intersection for all other cases and let the regex engine fail if needed

Does this sound correct to you @minrk?

Reply:
Awesome! I added come code now to handle that case, plus more tests. Let me know what you think!